[wasm] p/invoke and interop improvements

kg opened the pull request:
* Refactor logging in the wasm build tasks to use a LogAdapter helper class that can either write to stdout/stderr or via the msbuild logging APIs
* Refactor ManagedToNativeGenerator so that it can run in either in-process or out-of-process modes (but it still runs in-process right now). This will enable us to pivot to out-of-process in order to pick up function pointer support later on.
* Add support for function pointer types via reflection (so that we can still build and run on net4x for visual studio). Right now the underlying libraries are too old for this support to work, but once the libraries are updated it will.
* Add basic support for structs as arguments
* Implement scalar struct behavior from the wasm C ABI (structs containing one scalar are passed by-value instead of by-reference) in various places in the build tasks, interpreter, and AOT compiler
* Remove a use of varargs pinvoke from SystemNative because it isn't compatible with the WASM C ABI (it's replaced with two non-varargs pinvokes) 
* Generate more accurate C signatures for pinvokes

kg committed:
Checkpoint

Checkpoint

Checkpoint

Checkpoint

Update tests; make fn pointers explicitly blittable on net8

Checkpoint

Checkpoint

Update test assertions to handle out-of-process line breaking

Csproj no longer needed

Repair merge damage

Blittable structs

Fix broken WBT tests

Checkpoint wasm ABI

Wasm ABI in SignatureMapper

Wasm ABI fixes
Add test covering most of the weird parts of the wasm ABI (doesn't currently pass)

Checkpoint (this is broken)

Revert transform.c changes

Split test into both AOT and interp variants

Checkpoint

Checkpoint: scalarized calls seem to work now but the return value causes crashes

Update build_args_from_sig to properly support scalarized struct parameters on wasm

Fix scalarized struct return values in interp

Add missing file

kg commented:
cc @lewing

kg committed:
Repair merge damage

lewing committed:
Merge branch 'main' into pinvoke-srm-1

kg committed:
Run M2NG in-process again

kg committed:
Our current approach to SystemNative_SNPrintF is incorrect on some architectures, so change to an approach that always works

kg committed:
Support initonly fields in blittable structs in PInvokeTableGenerator etc

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg committed:
Work around our rampant use of warnaserror

kg committed:
Revert most csproj changes

kg committed:
Better solution for warnaserror (probably)

kg committed:
Restore old EnumCalendarInfo since we're in-process again

kg committed:
Revert fn ptr test changes

kg committed:
Maybe fix blazor WBTs

kg committed:
Make LogAdapter Info high priority so it appears in failed build logs

kg committed:
Hack to make skiasharp build

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg committed:
Implement wasm abi struct scalar rules in interp again

kg committed:
Add test coverage for i64 pinvokes + add fixme comments

kg committed:
Fix scalar int64 structs

radical commented on src/tasks/WasmAppBuilder/LogAdapter.cs:
```diff
@@ -0,0 +1,71 @@
+            Output?.WriteLine(AutoFormat(s, o));
+        }
+
+        public void LogMessage (MessageImportance mi, string s, params object[] o) {
```
Should this also drop `Log` prefix like the other methods?

kg replied:
I left it this way so existing code would keep working, I can go through and rename all the call sites if you want (it looks like there are about 30)

radical replied:
oh then I'm fine with either.

radical commented:
/azp run runtime-wasm

kg commented on src/tasks/WasmAppBuilder/LogAdapter.cs:
```diff
@@ -0,0 +1,71 @@
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+
+namespace WasmAppBuilder {
```
I did this originally as requested and it broke the build.

radical replied:
oh? We use that all over the place in tasks.

kg commented on src/tasks/WasmAppBuilder/LogAdapter.cs:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        public void Info (string code, string message, params object[] args) {
+            // We use MessageImportance.High to ensure this appears in build output, since
```
The warnings go nowhere otherwise, we need a way to pipe diagnostic information to the user, since we can't use Warning severity

radical replied:
Not sure I understand. Is that when running out-of-proc?

kg replied:
We build with warnaserror in a bunch of places, so the only way we can communicate information out of these build tasks without breaking the build is as high-importance messages, afaik

kg commented on src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs:
```diff
@@ -104,6 +137,119 @@ private void ExecuteInternal()
+            logAdapter.Error("WASM0066", $"WasmAppBuilder external process failed with exit code {exitCode}!");
+    }
+
+    internal readonly record struct ExecuteArguments (
```
can structs be sealed? you can't inherit from them

radical replied:
Oops!

kg commented on src/tasks/WasmAppBuilder/ManagedToNativeGenerator.cs:
```diff
@@ -66,33 +66,66 @@ public override bool Execute()
-                Assembly asm = mlc.LoadFromAssemblyPath(asmPath);
-                pinvoke.ScanAssembly(asm);
-                icall.ScanAssembly(asm);
+                // If we can't read a directory for any reason just skip it
```
This behavior is from the existing .NET code I was asked to duplicate, I don't know why it's there

kg commented on src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:
```diff
@@ -386,12 +419,94 @@ private bool HasAssemblyDisableRuntimeMarshallingAttribute(Assembly assembly)
+                return false;
+
+            if (!type.IsValueType) {
+                log.Info("WASM0060", "Type {0} is not blittable: Not a ValueType", type);
```
Can you provide an example of named warning codes? I'm unfamiliar with them

radical replied:
I just meant for example - `const string WarnNotABlittableType = "WASM0060"`

kg committed:
Code cleanup

kg committed:
Formatting

kg committed:
Checkpoint (reverse out-of-process)

kg committed:
Checkpoint

radical committed:
Merge branch 'main' into pinvoke-srm-1

radical commented:
/azp run runtime-wasm

radical commented:
There are a couple of failures which seem relevant, like:

[System.Runtime.Intrinsics.Tests aot](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-94446-merge-53f513650366430c87/normal-System.Runtime.Intrinsics.Tests/1/console.1a2c8876.log?helixlogtype=result):
```
[19:37:04] info: Starting:    System.Runtime.Intrinsics.Tests.dll
[19:37:05] info: /root/helix/work/workitem/e/wasm_build/AppBundle/_framework/dotnet.native.js:196: RuntimeError: null function or function signature mismatch
[19:37:05] info:       throw toThrow;
[19:37:05] info:       ^
[19:37:05] info: RuntimeError: null function or function signature mismatch
[19:37:05] info:     at aot_instances_System_Runtime_Intrinsics_Tests_Vectors_Vector64Tests_TestGetOne_uint16 (wasm://wasm/091f39e2:wasm-function[34037]:0x714a46)
[19:37:05] info:     at System_Runtime_Intrinsics_Tests_System_Runtime_Intrinsics_Tests_Vectors_Vector64Tests_GetOneUInt16 (wasm://wasm/091f39e2:wasm-function[23971]:0x497f53)
[19:37:05] info:     at aot_instances_aot_wrapper_gsharedvt_out_sig_void_this_ (wasm://wasm/091f39e2:wasm-function[29908]:0x5fda13)
[19:37:05] info:     at aot_instances_wrapper_runtime_invoke_object_runtime_invoke_sig_void_intptr_intptr_object_intptr_intptr_intptr (wasm://wasm/091f39e2:wasm-function[37318]:0x7e6ee8)
[19:37:05] info:     at mono_llvmonly_runtime_invoke (wasm://wasm/091f39e2:wasm-function[54970]:0xc48da7)
...
```

kg commented:
Interesting, it was passing tests before. Something must've changed in the BCL

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg committed:
Fix Vector64 AOT test failures (calling convention was incorrect when gsharedvt and simd types were involved)

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg committed:
Ensure we get MLC 8.0 in WasmAppBuilder

kg committed:
Update tests since function pointers work now

kg committed:
Attempt to resolve prebuilts/source build CI failure by adding (missing?) dependency on MLC
Explicitly track our target version of MLC
Bump our target versions of SRM and SCI

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

mthalman commented:
The source build error is:

```
.packages/microsoft.dotnet.arcade.sdk/9.0.0-beta.23607.2/tools/SourceBuild/AfterSourceBuild.proj(68,5): error : (NETCORE_ENGINEERING_TELEMETRY=AfterSourceBuild) 1 new pre-builts discovered! Detailed usage report can be found at /__w/1/s/artifacts/source-build/self/prebuilt-report/baseline-comparison.xml.
See https://aka.ms/dotnet/prebuilts for guidance on what pre-builts are and how to eliminate them.
Package IDs are:
System.Reflection.MetadataLoadContext.8.0.0
```

This will require adding System.Reflection.MetadataLoadContext.8.0.0 to [SBRP](https://github.com/dotnet/source-build-reference-packages?tab=readme-ov-file#adding-new-packages). Once that is merged, it will produce an SBRP package that will flow into runtime ([example](https://github.com/dotnet/runtime/pull/96382)). Once that is merged into runtime main, merge the latest main branch into this PR and it should resolve the prebuilt.

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg commented:
It looks like SBRP hasn't updated since Jan 5. Is there something I need to do to make the flow happen? Should I just manually do it in this PR? https://github.com/dotnet/runtime/pull/96989 happened, which I thought would have updated it to a more recent commit, but it didn't. @mthalman

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg commented:
@BrzVlad @kotlarmilos @vargaz Any of you feel comfortable reviewing the interp/type system changes here? They're nuanced enough I want to make sure someone with expertise looks them over.

kotlarmilos commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -4948,6 +4948,8 @@ process_call (EmitContext *ctx, MonoBasicBlock *bb, LLVMBuilderRef *builder_ref,
 		if (!addresses [call->inst.dreg])
 			addresses [call->inst.dreg] = build_alloca_address (ctx, sig->ret);
 		emit_store (builder, lcall, convert_full (ctx, addresses [call->inst.dreg]->value, pointer_type (LLVMTypeOf (lcall)), FALSE), is_volatile);
+		load_name = "wasm_vtype_as_scalar";
```
Should this be WASM only?

kotlarmilos reviewed (commented):
LGTM!

vargaz commented on src/mono/mono/mini/aot-runtime-wasm.c:
```diff
@@ -14,6 +14,9 @@
 #ifdef HOST_WASM
 
+gboolean
+mini_wasm_is_scalar_vtype (MonoType *type, MonoType **etype);
```
This is declared in mini-wasm.h.

vargaz commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -6154,8 +6157,15 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
+						/* SIMD value */
+						g_assert (lhs);
+						retval = LLVMBuildBitCast (builder, lhs, ret_type, "");
+						mono_llvm_dump_value (lhs);
```
Leftovers.

vargaz commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -1345,10 +1345,22 @@ typedef enum {
 
+static MonoType *
+filter_type_for_args_from_sig (MonoType *type) {
+#if defined(HOST_WASM) || defined(TARGET_WASM)
```
HOST_WASM should be enough.

kg committed:
Merge remote-tracking branch 'upstream/main' into pinvoke-srm-1

kg committed:
Address PR feedback

kg committed:
Repair merge damage
Address PR feedback

lewing commented on src/native/libs/System.Native/pal_string.h:
```diff
@@ -15,3 +15,12 @@
  */
 PALEXPORT int32_t SystemNative_SNPrintF(char* string, int32_t size, const char* format, ...);
+
+/**
+ * Two specialized overloads for use from Interop.Sys, because these two signatures are not equivalent
+ * on some architectures (like 64-bit WebAssembly)
+*/
+
+PALEXPORT int32_t SystemNative_SNPrintF_1S(char* string, int32_t size, const char* format, char* str);
+
+PALEXPORT int32_t SystemNative_SNPrintF_1I(char* string, int32_t size, const char* format, int arg);
```
@AaronRobinsonMSFT how did you handle this in runtimelab ?

AaronRobinsonMSFT replied:
This is the correct way to handle this scenario. Native varargs at the interop boundary have narrow support in .NET and are exclusive to the Windows platform and typically only for the C++/CLI scenario. See https://github.com/dotnet/runtime/issues/48796 for broader support.

The approach taken here is to provide an unmanaged export that managed code can call safely which then forwards to a native varargs function. This ensures the unmanaged calling convention is handled correctly during the P/Invoke and correctly at the dispatch to the unmanaged varargs.

It is ugly and not what anyone really wants, but the cost of supporting native varargs is expensive to support. This comes up rarely relative to other interop scenarios so just hasn't been prioritized. The aforementioned issue mentions some of the costs and relevant parties that would need to respond to support.

lewing replied:
Makes perfect sense, just wanted to verify.

lewing committed:
Merge branch 'main' into pinvoke-srm-1

lewing commented:
@kg please double check the conflict resolution with https://github.com/dotnet/runtime/pull/97643/files#r1483684742 in mind

kg commented:
> @kg please double check the conflict resolution with https://github.com/dotnet/runtime/pull/97643/files#r1483684742 in mind

Looks right to me, and tests should fail if it's wrong. Thanks for sorting out the merge

kg commented:
S.T.J failures are known, merging

am11 commented on eng/Versions.props:
```diff
@@ -123,16 +123,16 @@
     <!-- Keep toolset versions in sync with dotnet/msbuild and dotnet/sdk -->
-    <SystemReflectionMetadataToolsetVersion>7.0.0</SystemReflectionMetadataToolsetVersion>
-    <SystemReflectionMetadataLoadContextVersion>7.0.0</SystemReflectionMetadataLoadContextVersion>
+    <SystemReflectionMetadataToolsetVersion>8.0.0</SystemReflectionMetadataToolsetVersion>
```
Based on the comment above, this toolset version should not be updated without coordination with msbuild and sdk repos.

@ViktorHofer, FYI, this is blocking the codeflow https://github.com/dotnet/sdk/pull/38665 (while previous change made it to the next phases without problems https://github.com/dotnet/sdk/pull/38553). We can either revert this or update https://github.com/dotnet/sdk/blob/b32df9afff77e8f1afbd417e7c70841867cb6cdc/src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj#L75 to 8.0 (if we choose the latter, we can close https://github.com/dotnet/runtime/issues/96809).

kg replied:
I coordinated with people from the msbuild team at multiple points while developing this along with the SBRP team, so I'm sorry that something fell through the cracks here. Is there a specific step we missed?

kg replied:
I see now from looking through the replies to Larry's comment on this issue that you told us there. For some reason, GitHub did not send me a notification for your comment, so I didn't see it :( I guess in the future blocking issues like that need to be PR comments instead of replies.

lewing replied:
My preference would definitely be to take the bump in the sdk. VS 17.9 is the where these changes will ship and it has taken the update.

cc @ViktorHofer @marcpopMSFT

am11 replied:
NP, https://github.com/dotnet/runtime/issues/96809 is tracking it, but it seems (almost) resolved by your PR. Once confirmed, we can just update [sdk's Microsoft.NET.Build.Tasks.csproj#L74-L76](https://github.com/dotnet/sdk/blob/b32df9afff77e8f1afbd417e7c70841867cb6cdc/src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj#L74-L76) to 8.0 (now that all the related toolings are using 8.0; msbuild, HostModel and VS).

ViktorHofer replied:
I **think** for this to work, sdk would need to upgrade its .NET Framework leg to a VS 17.9 Preview image. I'm not sure if such is already available. Worst case, we would need to revert this dependency bump here. @lewing can you please make sure that this dependency flow issue gets resolved: https://github.com/dotnet/sdk/pull/38665?

akoeplinger replied:
Correct me if I'm wrong but the sdk _build_ should be fine, it'd just need the testing legs to use VS17.9 Preview right? That should already be the case: https://github.com/dotnet/sdk/blob/0962c1f89f5daf924a9fe876c80e80b0bde63b0d/.vsts-ci.yml#L108-L111

I'll try bumping the .csproj in the sdk PR and see what happens.

ViktorHofer replied:
Correct. Unification at run time.

akoeplinger replied:
Looks like it worked :)
